evm-types.md

```diff
@@ -423,11 +423,18 @@ A cons-list is used for the EVM wordstack.
     rule (W0 : WS):WordStack [ N := W ] => W0 : (WS [ N -Int 1 := W ]) requires N  >Int 0
 ```
 
--   Definedness conditions for `WS [ N ]` and `WS [ N := W ]`
+-   Definedness conditions for `WS [ N ]`, `WS [ N := W ]`, `BA [ N .. W ]`.
 
 ```{.k .symbolic}
-    rule #Ceil(WS[N])        => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS))) #Equals true}  [anywhere]
-    rule #Ceil(WS[ N := W ]) => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS))) #Equals true}  [anywhere]
+    rule #Ceil(WS[N])        => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS)))            #Equals true}  [anywhere]
+    rule #Ceil(WS[ N := W ]) => {((0 <=Int N) andBool (N <Int #sizeWordStack(WS)))            #Equals true}  [anywhere]
+    rule #Ceil(BA[ N .. W ]) => {((0 <=Int N) andBool (0 <=Int W))                            #Equals true}  [anywhere]
+    rule #Ceil(BA[ N := _:ByteArray ]) => {(0 <=Int N)                                        #Equals true}  [anywhere]
+    rule #Ceil( #range(LM, START, WIDTH) ) => {((0 <=Int START) andBool (0 <=Int WIDTH))      #Equals true}  [anywhere]
+    rule #Ceil(#padToWidth(N, _))      => {(0 <=Int N)                                        #Equals true}  [anywhere]
```

Comment:
I'm not sure this is correct.

I think that `#padToWidth` is, basically, defined as `padLeftBytes` (at least when working with `ByteArray`), and, in `domains.k`, `padLeftBytes` has two definitions: as a builtin and as a normal function. Interestingly enough, when it is defined as a normal function, it seems to be a total function. So, in the past, I wanted to mark it as a total function, but I think that @ehildenb explained to me that, being a builtin, whoever implements it can choose whether it is total or not. I hope I'm remembering this correctly.

Anyway, if that's indeed the case, then you can't define this `ceil` as you did since it may be different for different backends. Possible solutions include
```
#Ceil(#padToWidth(...))      => #Ceil(padLeftBytes(...))
```
Hopefully the Haskell backend would be able to solve that at some future point. Redefining `#padToWidth` to make it explicit that it's undefined when `N<0` would also work, I guess.

Reply:
I guess we need this rule now. Eventually will add a todo. What do you mean "Redefining #padToWidth to make it explicit that it's undefined when N<0 would also work" ? Isn't `#Ceil` the explicit way to define when function is undefined?

Reply:
Redefining `#padToWidth`:
```
    rule #padToWidth(N, BS)      => padLeftBytes(BS, N, 0) requires N >=Int 0  [concrete]
```

The main question is whether `#Ceil` should be in sync with the function definition rules and the builtin implementation. If the `ceil` axiom says that the function is defined iff `n>=0`, but the backend believes that the same function is total, is that reasonable? If `phi` is a formula, then `phi` should be equivalent to `phi and ceil(phi)`. So we could evaluate `#padToWidth` in two inconsistent ways:

```
#padToWidth(-1, B) = B  // evaluated as a builtin.

#padToWidth(-1, B) = #padToWidth(-1, B) and ceil(#padToWidth(-1, B)) = #padToWidth(-1, B) and bottom = bottom.
```

@traiansf @ttuegel in case they have comments.

Reply:
@virgil-serbanuta I incorporated your suggestions but also added these 2 #Ceil lemmas:
```
    rule #Ceil(padLeftBytes(_, N, _))       => {(0 <=Int N)                                   #Equals true}  [anywhere]
    rule #Ceil(padRightBytes(_, N, _))      => {(0 <=Int N)                                   #Equals true}  [anywhere]
```
We still need to resolve `#Ceil` formulas for these functions. Alternative would be to add them to `domains.k`